### PR TITLE
implementation of pi aggregation circuit

### PR DESCRIPTION
**Pull request by zhenfeizhang**
dependency: https://github.com/scroll-tech/zkevm-circuits/pull/519

### CODE REVIEW

File: aggregator/src/chunk.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+use ethers_core::utils::keccak256;
+
+#[derive(Default, Debug, Clone)]
+/// A chunk is a set of continuous blocks.
+/// A ChunkHash consists of 4 hashes, representing the changes incurred by this chunk of blocks:
+/// - state root before this chunk
+/// - state root after this chunk
+/// - the withdraw root of this chunk
+/// - the data hash of this chunk
+pub struct ChunkHash {
```
Comment (icemelon) [inline aggregator/src/chunk.rs:13]:
```suggestion
pub struct ChunkHash {
    pub(crate) chain_id: u8,
```

File: aggregator/src/batch.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        let preimage = chunk_hashes
+            .iter()
+            .flat_map(|chunk_hash| chunk_hash.data_hash.0.iter())
+            .cloned()
+            .collect::<Vec<_>>();
+        let data_hash = keccak256(preimage);
+
+        // public input hash is build as
+        //  keccak(
+        //      chain_id |
```
Comment (icemelon) [inline aggregator/src/batch.rs:36]: use `||` instead for concatenation?
```suggestion
        //      chain_id ||
```

File: aggregator/src/circuit.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            .cloned()
+            .collect();
+        res.push(batch_data_hash_preimage);
+
+        // compute piHash for each chunk for i in [0..k)
+        // chunk[i].piHash =
+        // keccak(chunk[i].prevStateRoot || chunk[i].postStateRoot || chunk[i].withdrawRoot ||
+        //        chunk[i].datahash)
+        for chunk in self.chunks.iter() {
+            let chunk_pi_hash_preimage = [
```
Comment (icemelon) [inline aggregator/src/circuit.rs:98]:
```suggestion
            let chunk_pi_hash_preimage = [
                self.chain_id,
```

File: aggregator/src/circuit.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+use super::{
+    batch::BatchHash,
+    chunk::ChunkHash,
+    config::{BatchCircuitConfig, BatchCircuitConfigArgs},
+};
+
+/// BatchCircuit
+#[derive(Clone, Debug, Default)]
+pub struct BatchHashCircuit<F: Field> {
```
Comment (icemelon) [inline aggregator/src/circuit.rs:19]:
```suggestion
pub struct BatchHashCircuit<F: Field> {
    pub(crate) chunk_id: u8,
```

File: aggregator/src/chunk.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    /// Chain identifier
+    pub(crate) chain_id: u8,
```
Comment (icemelon) [inline aggregator/src/chunk.rs:15]: sorry, made a mistake before. `chain_id` should be `u32` type instead of `u8`.

Comment (zhenfeizhang) [reply aggregator/src/chunk.rs]: fixed with a281f267fd00930e7322bfe47efec2e3a91a21bf

File: aggregator/src/config.rs [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+                        hash_input_cells.push(current_hash_input_cells);
+                        hash_output_cells.push(current_hash_output_cells);
+                        current_hash_input_cells = vec![];
+                        current_hash_output_cells = vec![];
+                    }
+                }
+
+                // sanity: we have same number of hash input and output
+                let hash_num = hash_input_cells.len();
+                let num_chunks = hash_num - 2;
```
Comment (huwenqing0606) [inline aggregator/src/config.rs:147]: Just to confirm: this -2 comes from the fact we take out `batch_public_input_hash_preimage` and `batch_data_hash_preimage`, and the rest are chunk hash preimages?

Comment (zhenfeizhang) [reply aggregator/src/config.rs:147]: yes that is correct

**Comment by zhenfeizhang** [comment]
closed in favor of https://github.com/scroll-tech/zkevm-circuits/pull/522

**Pull request closed without merging**